### ISSUE TITLE
Update query fails when entity has GUID as primary-key

### ISSUE THREAD

**Issue by baracuda69** [dotnet/efcore#7719]
Describe what is not working as expected.
Update query contains a select (after the update):
SELECT * 
FROM <TABLE> 
WHERE @@ROWCOUNT = 1 AND [Id] = scope_identity()

In the case where entity has GUID as the PK (as below), scope_identity() is NULL and, more importantly, int (numeric);

The EF-Core 1.1.0 generated SQL (for the update statement) looks like this:

exec sp_executesql N'SET NOCOUNT ON;
UPDATE [Leads] SET [Address1] = @p0, [Address2] = @p1, [ApproveReceiveCommercial] = @p2, [BirthDate] = @p3, [BusinessUnitId] = @p4, [CallbackStatus] = @p5, [City] = @p6, [CountryId] = @p7, [FirstName] = @p8, [IsArchived] = @p9, [IsImported] = @p10, [IsMale] = @p11, [LanguageCode] = @p12, [LastModifiedById] = @p13, [LastModifiedDate] = @p14, [LastName] = @p15, [ManagerAgentId] = @p16, [PhoneCountryCode] = @p17, [PhoneNumber] = @p18, [PhoneNumberAlternate] = @p19, [PoolId] = @p20, [Referral] = @p21, [RetentionAgentId] = @p22, [SalesAgentId] = @p23, [State] = @p24, [TitleId] = @p25, [UserId] = @p26, [UserName] = @p27, [ZipCode] = @p28
WHERE [Id] = @p29;

SELECT [Id], [CreatedDate]
FROM [Leads]
WHERE @@ROWCOUNT = 1 AND [Id] = scope_identity();

',N'@p29 uniqueidentifier,@p0 nvarchar(100),@p1 nvarchar(100),@p2 bit,@p3 datetime2(7),@p4 int,@p5 int,@p6 nvarchar(50),@p7 uniqueidentifier,@p8 nvarchar(50),@p9 bit,@p10 bit,@p11 bit,@p12 nvarchar(10),@p13 nvarchar(4000),@p14 datetime2(7),@p15 nvarchar(50),@p16 nvarchar(4000),@p17 nvarchar(10),@p18 nvarchar(50),@p19 nvarchar(50),@p20 uniqueidentifier,@p21 nvarchar(100),@p22 nvarchar(4000),@p23 uniqueidentifier,@p24 nvarchar(50),@p25 int,@p26 int,@p27 nvarchar(50),@p28 nvarchar(20)',@p29='49507C33-0E40-42E4-8F78-000F843F64AD',@p0=NULL,@p1=NULL,@p2=0,@p3='1970-01-01 00:00:00',@p4=1,@p5=3,@p6=N'',@p7='B239C10A-FDB0-DF11-AB2E-005056990011',@p8=N'test',@p9=0,@p10=1,@p11=1,@p12=N'ENGLISH',@p13=NULL,@p14=NULL,@p15=N'tester',@p16=NULL,@p17=N'+1',@p18=N'5551234',@p19=NULL,@p20='F5E33C99-E3C8-41D1-9449-690BAE5506F4',@p21=N'1',@p22=NULL,@p23='E7BAC065-5E46-4A85-B9D6-BD52BEE58118',@p24=NULL,@p25=NULL,@p26=NULL,@p27=N'tester@gmail.com',@p28=NULL
go

If you are seeing an exception, include the full exceptions details (message and stack trace).

Exception message: Operand type clash: uniqueidentifier is incompatible with numeric
Stack trace:
StackTrace Information
*********************************************
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlDataReader.TryConsumeMetaData()
   at System.Data.SqlClient.SqlDataReader.get_MetaData()
   at System.Data.SqlClient.SqlCommand.FinishExecuteReader(SqlDataReader ds, RunBehavior runBehavior, String resetOptionsString)
   at System.Data.SqlClient.SqlCommand.RunExecuteReaderTds(CommandBehavior cmdBehavior, RunBehavior runBehavior, Boolean returnStream, Boolean async, Int32 timeout, Task& task, Boolean asyncWrite, SqlDataReader ds)
   at System.Data.SqlClient.SqlCommand.ExecuteReader(CommandBehavior behavior)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, String executeMethod, IReadOnlyDictionary`2 parameterValues, Boolean closeConnection)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteReader(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)

### Steps to reproduce
Appears to be any entity that has PK value-type as GUID.

### Further technical details
EF Core version: 1.1.0
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows Server 2012 R2 (x64)
IDE: Visual Studio 2017RC

**Comment by ErikEJ** [comment]
Could you share your code/a repro project?

**Comment by baracuda69** [comment]
I think that I realise now the problem:

the POCO class "Lead" has the following definition for the PK (in .NET 4.51):

[Key]
[DatabaseGenerated(DatabaseGeneratedOption.Identity)]
public Guid Id { get; set; }

The problem is that, in CORE, I get an error when I start the APP:
ArgumentException: Identity value generation cannot be used for the property 'Id' on entity type 'Lead' because the property type is 'Guid'. Identity value generation can only be used with signed integer properties.

I had, therefore, changed the DatabaseGeneratedOptionAttribute to:
[DatabaseGenerated(DatabaseGeneratedOption.Computed]

I am fairly certain that it is THAT DataAnnotation Attribute causing the fault in the update statement.

So - what can I do to set the DatabaseGeneratedOption as Identity for a GUID PK?
I realise that this is probably not the correct channel for that - but any advice here would be greatly appreciated.

Many thanks

**Comment by baracuda69** [comment]
Indeed, it turns out that the changes from EF 6.x to 7.x (core) were confusing;
My project (like that of other devs) is being converted from .net 4.5 to .net core.

Guid identities (in .net 4.5) require an attribute (or fluid mapping) to set DatabaseGeneratedOption.Identity.

However, in EF CORE, this throws an exception as soon as you run the application:
ArgumentException: Identity value generation cannot be used for the property 'Id' on entity type 'MyType' because the property type is 'Guid'. Identity value generation can only be used with signed integer properties.

At first, I swapped the DatabaseGeneratedOption to Computed, which seemed to work ok.
But, when attempting to EDIT the entity, that is when I started getting the strange behaviour that I reported initially.

By removing the DatabaseGenerated attribute from Guid Identities (from all POCO models), the problem is now fixed.

For anyone interested in the full picture, I had to add fluent mappings instead (for each of the POCO models with Guid Identity) so that they use a different mechanism for auto-identity:

builder.Entity<T>.Property(p => p.Id).ValueGeneratedOnAdd()

I hope someone else finds this useful.
